### DISCUSSION TITLE
Another multi-row implementation worth reference

### DISCUSSION THREAD

**Discussion by Merci-chao** [onemen/TabMixPlus#364] [General]
Hello @onemen,

I have been using your Tab Mix Plus for many years. Among its features, the multi-row tab was the one that kept me using Waterfox. Until later, I found an alternative with userchrome.js, which allowed me to use the newer versions of Firefox. Recently, I decided to write my own version that supports dragging tabs animation in multi-row mode, and I tried to implement various new ideas. Soon after, it was great to hear that Tab Mix Plus made a comeback. After trying out the multi-row mode, I felt there was more room for improvement. I understand that Tab Mix Plus is a powerful extension, and the multi-row mode is just a small feature among many, but I still hope that my little creation can bring you some inspiration and ideas:
https://github.com/Merci-chao/userChrome.js

Thank you for your great work!

![themesupport](https://github.com/user-attachments/assets/2ddb57ac-4592-4fa6-a987-6a2f12d490bc)

![tabdragging](https://github.com/user-attachments/assets/b1483116-e2b4-4577-a1cf-9f90c09e1624)

![pinnedtabs](https://github.com/user-attachments/assets/3bbe756b-e97e-4c37-9fdd-d060eb3d7f90)

**Comment by onemen** [comment]
Hello @Merci-chao,

very nice work 👍 

If you want, we can work on some pull request together.

I want to check these features:
- Add tab-drag animation to multi-row. Remember that Tab Mix have feature to set tab with different width (width fit title).
- Maintain compatibility with Firefox feature to tab grouping, group tabs on drag. This is new Firefox feature, currently in Firefox nightly.
- Currently Tab mix offers options to control some context menu. In your `undoCloseTab` you implemented something similar but on tab-bar context menu. I would like to add the UI to Tab Mix, to enable users to control items in tab-bar context menu.

I would welcome anyone who would like to contribute.

**Comment by Merci-chao** [comment]
Thank you for your invitation. I'm sorry, I just need multiple rows for now and want to focus on my own script. If you need, I can help list some obvious bugs and areas that need improvement.

**Comment by onemen** [comment]
>  If you need, I can help list some obvious bugs and areas that need improvement.

I will appreciate any help.

**Comment by usrw** [comment]
Could there be any possibility to support dragging tabs animation in multi-row mode like the js mentioned above?

**Comment by Merci-chao** [comment]
@onemen I'm working on a new version of my script, and I've implemented a new animate mechanism that supports elements with variable widths - like tab groups and tabs that fit their title.

The idea is to avoid using CSS transform to simulate repositioning. Instead, the tabs are repositioned instantly during dragging (so the final layout is immediately known without manual calculations), and then the movement is "replayed" with animation.

The downside is that this approach triggers additional reflows, which can be performance-intensive.

[MultiTabRows@Merci.chao.uc.txt](https://github.com/user-attachments/files/21050049/MultiTabRows%40Merci.chao.uc.txt)
Here's the demo for your reference. It's not the final version and still has a few minor bugs.
I'd also suggest not digging too deep into my code, it's not well commented. 😅

**Reply by Merci-chao** [reply]
In addition, doing this thoroughly could get really complicated, so it’s better to keep it low on the priority list.

**Reply by onemen** [reply]
OK

**Comment by onemen** [comment]
@Merci-chao 

Did you see the new "pinned-tabs-container" Firefox will use from version 141
![image](https://github.com/user-attachments/assets/24fd1680-1f7a-4ad1-ae7c-5ccd9377abb6)

**Reply by Merci-chao** [reply]
Ya, I know. I've tried a grid layout for the `pinnedTabsContainer`, similar to:
![385683723-3bbe756b-e97e-4c37-9fdd-d060eb3d7f90](https://github.com/user-attachments/assets/4f57a267-e96e-44c0-b552-d41bceae323b)
but it causes a tricky problem in my case so I've just decided to move all tabs back to the `arrowScrollbox` and everything works great so far:
````javascript
let {pinnedTabsContainer} = gBrowser;
if (pinnedTabsContainer) {
	{
		let {contains} = pinnedTabsContainer;
		assign(pinnedTabsContainer, {
			appendChild: function(tab) {
				arrowScrollbox.insertBefore(tab, [...arrowScrollbox.children].find(t => !t.pinned));
			},
			insertBefore: function(...args) {
				arrowScrollbox.insertBefore(...args);
			},
			contains: function(node) {
				return new Error().stack.match(/^on_drop@/m) ?
					!!node?.closest("#tabbrowser-arrowscrollbox .tabbrowser-tab[pinned]") :
					contains.apply(this, arguments);
			},
		});
	}

	{
		let {contains} = arrowScrollbox;
		assign(arrowScrollbox, {
			prepend: function(...tabs) {
				let firstNormal = [...arrowScrollbox.children].find(t => !t.pinned && !tabs.includes(t));
				tabs.forEach(tab => arrowScrollbox.insertBefore(tab, firstNormal));
			},
			contains: function(node) {
				return new Error().stack.match(/^on_drop@/m) ?
					!!node?.closest("#tabbrowser-arrowscrollbox .tabbrowser-tab:not([pinned])") :
					contains.apply(this, arguments);
			},
		});
	}

	arrowScrollbox.prepend(...tabContainer.visibleTabs.slice(0, gBrowser.pinnedTabCount));
}
